### ISSUE TITLE
Documentation on sync protocol incorrectly states allowed_path_regex is a list of strings

### ISSUE THREAD

**Issue by jamiestrachan-shopify** [google/santa#933]
Hi there,
The [documentation here](https://santa.dev/development/sync-protocol.html) for the preflight sync protocol states that `allowed_path_regex` is of type `list of strings`, with the description "List of regular expressions to allow a binary to execute from a path".

![image](https://user-images.githubusercontent.com/98484497/200397434-e5a2b971-1e19-4f00-bf26-b3fd947259d5.png)

(This applies to `blocked_path_regex` as well.)

As far as I can tell, this field is a single string/pattern. Introducing the array notation prevents the feature from matching anything.
For example, setting the following will match the path "/opt/mytest/testbinary":
```
{
    "client_mode":"MONITOR",
    "clean_sync":false,
    "batch_size":50,
    "allowed_path_regex":"/opt/mytest/testbinary",
    etc...
}
```
However, the following configuration that introduces the list notation like in the example will _not_ match:
```
{
    "client_mode":"MONITOR",
    "clean_sync":false,
    "batch_size":50,
    "allowed_path_regex":["/opt/mytest/testbinary"],
    etc...
}
```

Here on [this line in the test](https://github.com/google/santa/blob/41c918ee87c32f41480b5ff922f016a3d964f535/Source/santasyncservice/SNTSyncTest.m#L512), we see a string, and not an array:
```
self.syncState.allowlistRegex = @"^horse$";
```

On [this line](https://github.com/google/santa/blob/41c918ee87c32f41480b5ff922f016a3d964f535/Source/santasyncservice/SNTSyncPreflight.m#L150
), we can see that the value set in the previous test is set directly from the preflight response, without converting from an array:  
```
self.syncState.allowlistRegex = resp[kAllowedPathRegex];
```

I would absolutely love to see this be a list of strings, but I am pretty sure it will be much easier to update the documentation!

**Comment by russellhancox** [comment]
Thanks for spotting this, I've opened a PR to fix the documentation

> I would absolutely love to see this be a list of strings, but I am pretty sure it will be much easier to update the documentation!

Unfortunately these are a single regex instead of a list for performance reasons, it's too easy to create a list of regexes that must be looped over and executed individually in a very hot path and it's almost always possible to craft a single regex that covers all the cases of a list (albeit the result is often a bit difficult to read)

**Comment by pmarkowsky** [comment]
If we want to do this in the future we might want to look into using RE2, and `RE2::Set` or `FilteredRE2` as both could speed up.

- `RE2::Set` -- would create a single automaton  so we'd only have to pass over it once.
- `FilteredRE2` -- would check for string atoms then only apply the regexp that might match.

**Comment by russellhancox** [comment]
We looked at using RE2 before (see https://github.com/google/santa/issues/29), it was abandoned because it doesn't support negative lookahead which is heavily used

**Comment by pmarkowsky** [comment]
Going to mark this as closed as the bug was fixed in #934. 

Please feel free to reopen if there's more we should cover.
